cpp/include/cugraph_c/graph_functions.h

```diff
@@ -0,0 +1,104 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph_c/array.h>
+#include <cugraph_c/graph.h>
+#include <cugraph_c/resource_handle.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief       Opaque vertex pair type
+ */
+typedef struct {
+  int32_t align_;
+} cugraph_vertex_pairs_t;
+
+/**
+ * @brief       Create vertex_pairs
+ *
+ * Input data will be shuffled to the proper GPU and stored in the
+ * output vertex_pairs.
+ *
+ * @param [in]  handle       Handle for accessing resources
+ * @param [in]  graph        Graph to operate on
+ * @param [in]  first        Type erased array of vertex ids for the first vertex of the pair
+ * @param [in]  second       Type erased array of vertex ids for the second vertex of the pair
+ * @param [out] vertex_pairs Opaque pointer to vertex_pairs
+ * @param [out] error        Pointer to an error object storing details of any error.  Will
+ *                           be populated if error code is not CUGRAPH_SUCCESS
+ * @return error code
+ */
+cugraph_error_code_t cugraph_create_vertex_pairs(
+  const cugraph_resource_handle_t* handle,
+  cugraph_graph_t* graph,
+  const cugraph_type_erased_device_array_view_t* first,
+  const cugraph_type_erased_device_array_view_t* second,
+  cugraph_vertex_pairs_t** vertex_pairs,
+  cugraph_error_t** error);
+
+/**
+ * @brief       Get the first vertex id array
+ *
+ * @param [in]     vertex_pairs   A vertex_pairs
+ * @return type erased array of vertex ids
+ */
+cugraph_type_erased_device_array_view_t* cugraph_vertex_pairs_get_first(
+  cugraph_vertex_pairs_t* vertex_pairs);
+
+/**
+ * @brief       Get the second vertex id array
+ *
+ * @param [in]     vertex_pairs   A vertex_pairs
+ * @return type erased array of vertex ids
+ */
+cugraph_type_erased_device_array_view_t* cugraph_vertex_pairs_get_second(
+  cugraph_vertex_pairs_t* vertex_pairs);
+
+/**
+ * @brief     Free vertex pair
+ *
+ * @param [in]    vertex_pairs The vertex pairs
+ */
+void cugraph_vertex_pairs_free(cugraph_vertex_pairs_t* vertex_pairs);
+
+/**
+ * @brief      Find all 2-hop neighbors in the graph
+ *
+ * @param [in]  handle         Handle for accessing resources
+ * @param [in]  graph          Pointer to graph
+ * @param [in]  start_vertices Optional type erased array of starting vertices
+ *                             If NULL use all, if specified compute two-hop
+ *                             neighbors for these starting vertices
+ * @param [out] vertex_pairs   Opaque pointer to resulting vertex pairs
+ * @param [out] error          Pointer to an error object storing details of any error.  Will
+ *                             be populated if error code is not CUGRAPH_SUCCESS
+ * @return error code
+ */
+cugraph_error_code_t cugraph_two_hop_neighbors(
```

Comment:
I am having a concern as I am trying to expose this function not just to `pylibcugraph` but also [cugraph](https://github.com/rapidsai/cugraph/blob/branch-22.10/python/cugraph/cugraph/structure/graph_implementation/simpleGraph.py#L505). in python `cugraph`, this function will returns `first` and `second`(in fact, `pylibcugraoh` `get_two_hop_neighbors` will call `cugraph_create_vertex_pairs` which is fine. But the similarity algos needs pylibcugraph `get_two_hop_neighbors` to return `cugraph_vertex_pairs_t*` which will there be freed in the similarity algo. However, I think this can cause a leak if the user doesn't call the similarity algos. I think for now I can't make these two calls modulable(seperate) and call `cugraph_two_hop_neighbors` within the similarity algos. any thoughts?